chips/rp2040/src/timer.rs

```diff
@@ -0,0 +1,266 @@
+    VALUE OFFSET (0) NUMBITS (32) []
+],
+TIMERAWL [
+    VALUE OFFSET (0) NUMBITS (32) []
+],
+DBGPAUSE [
+    /// Pause when processor 1 is in debug mode
+    DBG1 OFFSET(2) NUMBITS(1) [],
+    /// Pause when processor 0 is in debug mode
+    DBG0 OFFSET(1) NUMBITS(1) []
+],
+PAUSE [
+
+    PAUSE OFFSET(0) NUMBITS(1) []
+],
+INTR [
+
+    ALARM_3 OFFSET(3) NUMBITS(1) [],
+
+    ALARM_2 OFFSET(2) NUMBITS(1) [],
+
+    ALARM_1 OFFSET(1) NUMBITS(1) [],
+
+    ALARM_0 OFFSET(0) NUMBITS(1) []
+],
+INTE [
+
+    ALARM_3 OFFSET(3) NUMBITS(1) [],
+
+    ALARM_2 OFFSET(2) NUMBITS(1) [],
+
+    ALARM_1 OFFSET(1) NUMBITS(1) [],
+
+    ALARM_0 OFFSET(0) NUMBITS(1) []
+],
+INTF [
+
+    ALARM_3 OFFSET(3) NUMBITS(1) [],
+
+    ALARM_2 OFFSET(2) NUMBITS(1) [],
+
+    ALARM_1 OFFSET(1) NUMBITS(1) [],
+
+    ALARM_0 OFFSET(0) NUMBITS(1) []
+],
+INTS [
+
+    ALARM_3 OFFSET(3) NUMBITS(1) [],
+
+    ALARM_2 OFFSET(2) NUMBITS(1) [],
+
+    ALARM_1 OFFSET(1) NUMBITS(1) [],
+
+    ALARM_0 OFFSET(0) NUMBITS(1) []
+]
+];
+const TIMER_BASE: StaticRef<TimerRegisters> =
+    unsafe { StaticRef::new(0x40054000 as *const TimerRegisters) };
+
+pub struct RPTimer<'a> {
+    registers: StaticRef<TimerRegisters>,
+    client: OptionalCell<&'a dyn hil::time::AlarmClient>,
+}
+
+impl<'a> RPTimer<'a> {
+    pub const fn new() -> RPTimer<'a> {
+        RPTimer {
+            registers: TIMER_BASE,
+            client: OptionalCell::empty(),
+        }
+    }
+
+    fn enable_interrupt(&self) {
+        self.registers.inte.modify(INTE::ALARM_0::SET);
+    }
+
+    fn disable_interrupt(&self) {
+        self.registers.inte.modify(INTE::ALARM_0::CLEAR);
+    }
+
+    fn enable_timer_interrupt(&self) {
+        unsafe {
+            atomic(|| {
+                let n = cortexm0p::nvic::Nvic::new(TIMER_IRQ_0);
+                n.enable();
+            })
+        }
+    }
+
+    fn disable_timer_interrupt(&self) {
+        unsafe {
+            cortexm0p::nvic::Nvic::new(TIMER_IRQ_0).disable();
+        }
+    }
+
+    pub fn handle_interrupt(&self) {
+        self.registers.intr.modify(INTR::ALARM_0::SET);
+        self.client.map(|client| client.alarm());
+    }
+}
+
+impl Time for RPTimer<'_> {
+    type Frequency = hil::time::Freq1MHz;
+    type Ticks = Ticks32;
+
+    fn now(&self) -> Self::Ticks {
+        Self::Ticks::from(self.registers.timerawl.get())
+    }
+}
+
+impl<'a> Alarm<'a> for RPTimer<'a> {
+    fn set_alarm_client(&self, client: &'a dyn hil::time::AlarmClient) {
+        self.client.set(client);
+    }
+
+    fn set_alarm(&self, reference: Self::Ticks, dt: Self::Ticks) {
+        let mut expire = reference.wrapping_add(dt);
+        let now = self.now();
+        if !now.within_range(reference, expire) {
+            expire = now;
+        }
+
+        if expire.wrapping_sub(now) < self.minimum_dt() {
+            expire = now.wrapping_add(self.minimum_dt());
+        }
+
+        self.registers.alarm0.set(expire.into_u32());
+        self.enable_timer_interrupt();
```

Comment:
why is this needed? Is the other `enable_interrupt` insufficient?

Reply:
From our tests that does not seem to be enough, I'll dig a little further and get back.

Reply:
We tested it and it does not work without it. I have commented this in the source.